### ISSUE TITLE
torch.mean returns a wildly incorrect result 0.3277 on YCbCr version of CIFAR10 on CPU with dtype=float32

### ISSUE THREAD

**Issue by philip-bl** [pytorch/pytorch#38716]
## 🐛 Bug

When I load CIFAR10 training dataset, convert it to YCbCr, convert it to tensors, transpose it to HWC format and turn it into one large CPU float32 tensor and calculate its channels means, the result is wildly incorrect.

Also, the incorrect result it produces is (0.3277, 0.3277, 0.3277) - the same as reported in https://github.com/pytorch/pytorch/issues/10248, where the author doesn't even convert it to YCbCr. I think this suggests that this is just some bug in `torch.mean`. Also, on my local machine I only encounter this bug if I try to calculate mean for all 3 channels at once. If I instead calculate it for only one channel, I don't get 0.3277.

## To Reproduce

Code snippet:

```python
from typing import Tuple
from PIL.Image import Image
from einops import rearrange # pip install einops
import torch
from torchvision.datasets import CIFAR10
from torchvision.transforms.functional import to_tensor

tv_ds = CIFAR10(".", download=True)
YCbCr_images: Tuple[Image, ...] = tuple(image.convert("YCbCr") for image, label in tv_ds)
tensors: Tuple[torch.Tensor, ...] = tuple(
    rearrange(to_tensor(image), "c h w -> h w c") for image in YCbCr_images
)
x = torch.stack(tensors).unsqueeze(0)  # 1 × 50000 × 32 × 32 × 3                                                                                                                                                                             
x_flattened = x.reshape(-1, 3)
print(x_flattened.mean(dim=0))  # (0.3277, 0.3277, 0.3277)                                                                                                                                                                                   
for channel in range(3):
    print(x_flattened[:, channel].mean())  # 0.3277, 0.3277, 0.3277 on google colab, but 0.4794, 0.4916, 0.5037 on my local machine                                                                                                                                                                     
print(x_flattened.double().mean(dim=0))  # (0.4790, 0.4809, 0.5077)                                                                                                                                                                          
print(x_flattened.cuda().mean(dim=0))  # (0.4790, 0.4809, 0.5077)
```

[Google colab link](https://colab.research.google.com/drive/1hV54M8zH3Cpx6n-A-JHSSuCVnZ1oBvdC?usp=sharing)

## Expected behavior

I expect to get the correct result rather than (0.3277, 0.3277, 0.3277).

## Environment

On my local machine:

```
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Debian GNU/Linux 10 (buster)
GCC version: (Debian 8.3.0-6) 8.3.0
CMake version: version 3.13.4

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 9.2.148
GPU models and configuration: 
GPU 0: GeForce RTX 2080 SUPER
GPU 1: GeForce GTX 1070

Nvidia driver version: 418.113
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.16.2
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.0                      166    conda-forge
[conda] numpy                     1.18.1           py38h95a1406_0    conda-forge
[conda] pytorch                   1.4.0           py3.8_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] pytorch-ignite            0.3.0                    pypi_0    pypi
[conda] pytorch-lightning         0.7.1                    pypi_0    pypi
[conda] torchvision               0.5.0                py38_cu101    pytorch
```

On google colab:

```
Collecting environment information...
PyTorch version: 1.5.0+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla K80
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.4
[pip3] torch==1.5.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.6.0+cu101
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @nairbv

**Comment by glaringlee** [comment]
mark high priority since mean() produced a potentially wrong number.

**Comment by emcastillo** [comment]
If I convert the data to a numpy array and calculate the means it yields the following
This is for torch 1.4.0
```
numpy [0.32768 0.32768 0.32768]
numpy channel 0 0.47898987
numpy channel 1 0.48092735
numpy channel 2 0.5077502

torch tensor([0.3277, 0.3277, 0.3277]) 
torch channel 0 tensor(0.4791)
torch channel 1 tensor(0.4810)
torch channel 2 tensor(0.5076)
torch double tensor([0.4790, 0.4809, 0.5077], dtype=torch.float64)
torch cuda tensor([0.4790, 0.4809, 0.5077], device='cuda:0')
numpy [0.32768 0.32768 0.32768]
```

This is with the following code

```python
from typing import Tuple
from PIL.Image import Image
from einops import rearrange  # pip install einops
import torch
from torchvision.datasets import CIFAR10
from torchvision.transforms.functional import to_tensor

tv_ds = CIFAR10(".", download=True)
YCbCr_images: Tuple[Image, ...] = tuple(
    image.convert("YCbCr") for image, label in tv_ds
)
tensors: Tuple[torch.Tensor, ...] = tuple(
    rearrange(to_tensor(image), "c h w -> h w c") for image in YCbCr_images
)
x = torch.stack(tensors).unsqueeze(0)  # 1 × 50000 × 32 × 32 × 3
x_flattened = x.reshape(-1, 3)
x_numpy = x_flattened.detach().numpy()
print("numpy", x_numpy.mean(axis=0))
for channel in range(3):
    print("numpy channel", channel, x_numpy[:, channel].mean())

print("torch", x_flattened.mean(dim=0))  # (0.3277, 0.3277, 0.3277)
for channel in range(3):
    print("torch channel", channel, x_flattened[:, channel].mean())
print("torch double", x_flattened.double().mean(dim=0))
print("torch cuda", x_flattened.cuda().mean(dim=0)) 
print("numpy", x_numpy.mean(axis=0)) 
```

Then if I change the last line to
```
print("numpy", x_flattened.double().detach().numpy().mean(axis=0)) 
```
I get the following result
```
numpy [0.32768 0.32768 0.32768]
numpy channel 0 0.47898987
numpy channel 1 0.48092735
numpy channel 2 0.5077502
torch tensor([0.3277, 0.3277, 0.3277])
torch channel 0 tensor(0.4791)
torch channel 1 tensor(0.4810)
torch channel 2 tensor(0.5076)
torch double tensor([0.4790, 0.4809, 0.5077], dtype=torch.float64)
torch cuda tensor([0.4790, 0.4809, 0.5077], device='cuda:0')
numpy [0.47899024 0.48092787 0.50774947]
```

Seems like the use comes from the tensor itself instead of `mean`?

**Comment by peterbell10** [comment]
The problem is that it calculates a running sum
```
sum = 0
for x in xs:
	sum += x
```

When `sum` is large enough, `x` is just a rounding error by comparison and so `sum += x` doesn't increase the `sum`. CUDA does a similar sum, but only for partial sums which are calculated in parallel before being added together. So, instead of `big + small = big` it's doing `medium + small = medium`, followed by `medium + medium = big`. The same is true for local machine vs colab. My local machine has more CPU cores, so `binary_kernel_reduce_vec` splits the sum into more intermediate partial sums.

Note that `double` should have the exact same precision issues, it will systematically under-report the sum and mean of large tensors. It will just only effect the lower mantissa bits.

One way to fix this would be to avoid calculating the sum and instead use the [`iterative mean`](http://www.heikohoffmann.de/htmlthesis/node134.html). This keeps the accumulator on the same order of magnitude as the data, so the addition doesn't lose precision. Unfortunately, it is also much slower because it does many more divisions and also it can't be vectorized in the current `Reduce.h` machinery.

I think it makes sense to instead focus on the accuracy of the `sum`. @emcastillo's suggestion of using a `double` for the accumulator should help here. However, that would also have performance implications as well as leaving the `double` precision sum inaccurate. 

I think a better option would be to split the reduction into more partial sums, similar to how it's done on GPU. That should have minimal performance implications and will also fix `double` tensors in exactly the same way.

Thoughts?

**Comment by gchanan** [comment]
removing 1.5.1 milestone based on @peterbell10's comments; at a minimum, it's not a trivial fix.

**Comment by emcastillo** [comment]
Exactly, I explored the idea of modifying the cpu reduction algorithm to do partial reductions instead of just the running sum. But I wasn't sure of the implications for other routines.
I can keep working in this direction if you think it is reasonable @peterbell10.

**Comment by ezyang** [comment]
cc @ngimel

**Comment by peterbell10** [comment]
I don't think it makes sense to modify the generic reduction algorithm. Most other users of `binary_kernel_reduce_vec` don't really have this problem: `and`, `or`, `min`, `max`.

`prod` would depend on the input; if all elements are magnitude > 1 then it definitely doesn't help. For some mixtures of > 1 and < 1 then it could prevent `inf`s popping up.

**Comment by ngimel** [comment]
This is a similar issue https://github.com/pytorch/pytorch/issues/37234 where the difference in results is attributable to contiguous vs discontiguous tensors (one is using vectorized kernel, another does not) and different number of cores on the machines. 
I agree that making `sum` accumulate in double has performance implications. `cumcum` for floats accumulates in double on cpu, and `cumcum` on floats is slower than cumsum on doubles.

**Comment by emcastillo** [comment]
I think that cuda splitting the sums in multiple partial sums comes naturally from how the reduction algorithm is implemented. In the CPU should we try to do the reduction manually for the sum without using the binary_kernel_reduce_vec template?

Also, how should we decide the grain in which we split the partial sums? I guess that a number similar to the block size of the gpu should suffice.

**Comment by emcastillo** [comment]
I have been tracing the execution of the sum reduction and this is what I found.

First, if instead of doing 
```python
a = numpy.random.rand(51200000, 3)
a = torch.tensor(a, dtype=torch.float32)
b = a.sum(axis=0)
print(b)
```
I do 
```python
a = numpy.random.rand(3, 51200000)
a = torch.tensor(a, dtype=torch.float32)
b = a.sum(axis=1)
print(b)
```
Then the result is more or less correct (float32 rounding errors still happening but now is expected).
The difference here is that the reduction is now contiguous and before it was outer in `binary_kernel_reduce_vec`, so this issue seems to be the same that @ngimel pointed out.

I will keep checking the outer reduction and updating this post.

**Comment by emcastillo** [comment]
Just pushed a PR that batches the reductions for cpu
#39512

**Comment by peterbell10** [comment]
It seems we've doubled up a bit here. See #39516 for an alternate solution that doesn't require a batch size and has good error performance for any length of tensor.
